get_fm_data.py

```diff
@@ -95,6 +117,174 @@ def log_failed_response(self, res, endpoint: str):
+        new_state = "Costs collected at " + now.isoformat()
+        result = {}
+        result['records'] = charging_cost_points
+        self.set_state("input_text.charging_costs", state=new_state, attributes=result)
+        self.set_value("input_number.total_charging_cost_last_7_days", total_charging_cost_last_7_days)
+
+    def get_charged_energy(self, *args, **kwargs):
+        """ Communicate with FM server and check the results.
+
+        Request charging columes of last 7 days from the server.
+        ToDo: make this period a setting for the user.
+        Make totals of charging and dis-charging per day and over the period
+
+        """
+        now = self.get_now()
+        # Getting data since start of yesterday so that user can look back a little further than just current window.
+        startDataPeriod = str((now + timedelta(days=-7)).date())
+        endDataPeriod = str((now + timedelta(days=-1)).date())
+
+        url_params = {
+            "event_starts_after": startDataPeriod + "T00:00:00.000Z",
+            "event_ends_before": endDataPeriod + "T23:59:59.999Z",
+        }
+        self.log(f"url_params: {url_params}")
+
+        res = requests.get(
+            self.CHARGE_POWER_URL,
+            params=url_params,
+            headers={"Authorization": self.fm_token},
+        )
+
+        # Authorisation error, retry authoristion.
+        if res.status_code == 401:
+            self.log_failed_response(res, "Get FM CHARGE POWER")
+            self.try_solve_authentication_error(res, self.CHARGE_POWER_URL, self.get_charging_energy, *args, **kwargs)
+            return
+
+        if res.status_code != 200:
+            self.log_failed_response(res, "Get FM CHARGE POWER")
+            # Currently there is no reason to retry as the server will not re-run scheduled script for cost calculation
+        charge_power_points = res.json()
+
+        total_charged_energy_last_7_days = 0
+        total_discharged_energy_last_7_days = 0
+        total_emissions_last_7_days = 0
+        total_saved_emissions_last_7_days = 0
+        total_minutes_charged = 0
+        total_minutes_discharged = 0
+        charging_energy_points = {}
+        charging_energy_points.clear
+        resolution_in_miliseconds = c.FM_EVENT_RESOLUTION_IN_MINUTES * 60 * 1000
+
+        for charge_power in charge_power_points:
+            # The API returns both actual and schedualed power, ignore the values from the scheduals
+            if charge_power['source']['type'] == "scheduler":
+                continue
+
+            power = float(charge_power['event_value'])
+            key = charge_power['event_start']
+            charging_energy_points[key] = power
+            if power is None:
+                continue
+
+            # Look up the emission mathcing with power['event_start'], this will be a match every 3 items
+            # as emission has a resolution of 15 minutes and power of 5 min.
```

Comment:
Any other combination of sensor resolutions will lead to unexpected results. Best raise an error explicitly in case of incompatible resolutions.

Reply:
Well, the power and emission response both do not contain an explicit resolution.
An average-time-between-events could be calculated. For emissions this seems fine but for power there might (are!) also events be missing.
The code indeed assumes an exact 15 min. interval in emissions and an exact match with the power event start. I don't have a "solid" way to handle let's say a 3 or 6 min. resolution. This BTW would have a big impact on the code in general..